sources/api/apiclient/src/exec.rs

```diff
@@ -0,0 +1,683 @@
+                        Err(e).context(error::ReadFromUser)?
+                    }
+                }
+            }
+        }
+    }
+
+    /// Read from stdin in bulk, sending larger batches of data at a time.
+    fn read_stdin(tx: mpsc::UnboundedSender<Message>, capacity: Arc<AtomicCapacity>) -> Result<()> {
+        let mut stdin = std::io::stdin();
+        // Keep track of the number of messages we've read.  We compare this to the number of
+        // messages the server has written, as received in its regular capacity update messages, so
+        // that we don't overwhelm the server.
+        let mut messages_read = 0u64;
+
+        loop {
+            // Wait for server to have capacity for writes before reading; don't give "false hope" to
+            // whatever's writing to our stdin that it'll be read until there's room for it.
+            Self::wait_for_capacity(messages_read, &capacity);
+
+            // Read a batch of data at a time; 4k is a balanced number for small and large jobs.
+            let mut buf = [0; 4096];
+            let count = stdin.retry_read(&mut buf).context(error::ReadFromUser)?;
+            // A read of 0 indicates EOF, so we're done.
+            if count == 0 {
+                break;
+            }
+            messages_read += 1;
+
+            // Send the data to the server in a Binary message without encoding.
+            let msg = Vec::from(&buf[..count]);
+            tx.unbounded_send(Message::Binary(msg))
+                .context(error::SendMessage { kind: "user input" })?;
+        }
+        debug!("Finished reading input");
+
+        // Send a ContentComplete message to the server so it can exit the process more cleanly.
+        // This is more important than the TTY case; interactive use typically has users typing
+        // exit, or quit, or ctrl-d... noninteractive programs typically wait for EOF.
+        let msg =
+            serde_json::to_string(&ClientMessage::ContentComplete).context(error::Serialize)?;
+        tx.unbounded_send(Message::Text(msg))
+            .context(error::SendMessage {
+                kind: "content complete",
+            })?;
+
+        Ok(())
+    }
+
+    /// Sleeps until the server has capacity to receive more process input.
+    ///
+    /// We know how many messages we've read from user input, and the AtomicCapacity is updated any
+    /// time the server sends us a capacity update.  We compare read count to written count to know
+    /// how many messages the server has yet to write, and if that's over the maximum number of
+    /// messages the server wants outstanding, we wait.  (The server will terminate us otherwise.)
+    fn wait_for_capacity(messages_read: u64, capacity: &Arc<AtomicCapacity>) {
+        let mut waited = 0u64;
+        loop {
+            let messages_outstanding =
+                messages_read - capacity.messages_written.load(Ordering::SeqCst);
```

Comment:
I assume in that failure case we'd terminate the client?

Reply:
If we make that function return a Result, what would naturally happen is that the stdin_fn thread would stop and the user would see an error, but the server wouldn't know, and overall we'd continue.  We could also send a ContentComplete in that case, but that's a bit misleading; that message is meant for a healthy case of the user being done.  I could change ReadFromUser such that it makes another channel that the main loop reads from and ends early...  I guess that's related to your other comment, in that this would be the first case we have where we'd want to end it early.  This should never happen, though!